### ISSUE TITLE
Is sorting supposed to occur when clicking column header in SQL tab ?

### ISSUE THREAD

**Issue by Bluezy** [sqlitebrowser/sqlitebrowser#1486]
What is your question
-----------------------
In the 'Execute SQL' tab, no sorting takes places when clicking on a column header. 
Isn't that what is supposed to happen, as it occurs in the 'Browse Data' tab ?
I seem to remember it worked like that in previous versions.

(running version 3.10.99 under Windows 10 X64)
Thanks for lighting my lantern...

**Comment by karim** [comment]
> Is sorting supposed to occur when clicking column header in SQL tab ?

Short answer: No. :smiley: 

What you do in "Browse Data" tab, from filtering columns to sorting, happens after executing extra SQL statements. However, "Execute SQL" simply execute your statements *as-is* without modifying them. If you want to sort the result you can use `ORDER BY` in your SQL statement.

> I seem to remember it worked like that in previous versions.

I rarely use the "Execute SQL" tab so I'm not really sure. If that was the case then I'm sorry, and I'm also surprised why it was removed. I also checked our latest stable version (3.10.1) that was released almost a year ago and it does not allow sorting columns in the "Execute SQL" tab, so I'm not sure which version you might be referring to. It must have been a very old version.

**Comment by MKleusberg** [comment]
I'm pretty sure this never worked in older versions. Maybe sorting in really old versions but definitely not filtering. As @karim said we just execute the SQL statement as-is and you would have to add a ```WHERE``` clause for filtering or a ```ORDER BY``` clause for sorting.

We could add this in theory in two ways:
1) Modifying the SQL statement. That's really, really complicated to do however, especially for more complex queries, because we would have to understand the entire query before modifying it.
2) By just sorting and/or filtering the table view and not querying the database again. That's much easier but doesn't really work either because of the lazy population mechanism we have. Also it's a performance problem for larger results.

And both approaches might be confusing for the user too because then the results and the executed statement don't match anymore.

So yeah, probably not going to happen anytime soon. Sorry :frowning:

**Comment by chrisjlocke** [comment]
Could the columns just not be sortable like the 'Browse Data' grid?  That only shows a limited amount of data, and works OK.  So long as in the Wiki its documented its limitations...  many times I've run a query and while I can amend it, clicking a column heading is far quicker, and to be honest, more in line with what the user expects - especially if they've used a similar grid only two tabs away...

**Comment by chrisjlocke** [comment]
> And both approaches might be confusing for the user too because then the results and the executed statement don't match anymore.

Personally, I think its obvious to the user whats happened and why the data sorted in the grid doesn't match the query.  As soon as they rerun the query, the sort order will match the query again, so the confusion will be 'short lived'.....
Just a thought.

**Comment by sky5walk** [comment]
Curious why the execute SQL query results cannot be placed within another data grid similar to the Browse Data tab?
Is there a performance hit with just duplicating the relevant code behind the data grid?

**Comment by justinclift** [comment]
> Curious why the execute SQL query results cannot be placed within another data grid similar to the Browse Data tab?

Sounds like a decent idea to me.  From memory - though I've not tried it myself - there is some capability for doing that already.  Something along the lines of saving the output as a view, and then being able to view the err... view directly in the Browse Data tab. :smile:

**Comment by MKleusberg** [comment]
@mgrojo I don't see the benefit of that to be honest. If we want to create a view and modify that we can just as well modify the select statement directly and re-execute that. Both are equally difficult for complex SQL statements. Or am I missing something?

@chrisjlocke You're probably right. And when in doubt we can also add some icon or so to indicate that the view was changed from the original statement.

@sky5walk The grids in the Browse Data tab and in the Execute SQL tabs are the same. It's just that in the Browse Data tab we detect clicks on the column headers and modify the underlying SQL statements accordingly, then refresh the entire view. But we can only do that because we control the SQL statements. For the Execute SQL tab where any statement is allowed this is much more complicated. So the only chance we have at the moment is to implement it in a different way by sorting it in the application, not in the database.

**Comment by chrisjlocke** [comment]
> we detect clicks on the column headers and modify the underlying SQL statements 

Sorry - just to confirm.  The grid doesn't have 'native' sorting built in?

**Comment by MKleusberg** [comment]
By "native" you mean in the grid? If so, then no, it doesn't. What you can do is putting a proxy model between the SQL model (which gets the data) and the grid (which displays the data), the proxy model can then sort the data after it is fetched from the database and before it is sent to the grid.

Why we don't use this in the Browse Data tab is performance in general, the prefetch mechanism, and the fact that sorting this way doesn't take any database specifics like collations into account.

**Comment by sky5walk** [comment]
Yes, I see the 'order by's in the SQL log.
I see 2 immediate options to avoid the parsing nightmare for complex SQL queries(assigned aliases, attach tables, etc).
1. Dump Execute SQL output to flat tables(SQL1,SQL2,etc.) and spawn the same processes used in the Browse data tab.
2. Add rudimentary grid like features for sorting and comparisons.

I prefer option 1 since that reuses existing code and allows additional queries on the newly created temp tables.

**Comment by mgrojo** [comment]
>  @mgrojo I don't see the benefit of that to be honest. If we want to create a view and modify that we can just as well modify the select statement directly and re-execute that. Both are equally difficult for complex SQL statements. Or am I missing something?

Maybe it's me who is missing something but what I can see is that we create views from very complex queries (it's only a matter of prepending CREATE VIEW foo AS...). And then we can use these views on the Browse Table with all the feature pack (except editing, in principle).

For example, given this example from https://www.sqlite.org/lang_with.html
```sql
CREATE TABLE org(
  name TEXT PRIMARY KEY,
  boss TEXT REFERENCES org
) WITHOUT ROWID;
INSERT INTO org VALUES('Alice',NULL);
INSERT INTO org VALUES('Bob','Alice');
INSERT INTO org VALUES('Cindy','Alice');
INSERT INTO org VALUES('Dave','Bob');
INSERT INTO org VALUES('Emma','Bob');
INSERT INTO org VALUES('Fred','Cindy');
INSERT INTO org VALUES('Gail','Cindy');

WITH RECURSIVE
  under_alice(name,level) AS (
    VALUES('Alice',0)
    UNION ALL
    SELECT org.name, under_alice.level+1
      FROM org JOIN under_alice ON org.boss=under_alice.name
     ORDER BY 2
  )
SELECT substr('..........',1,level*3) || name FROM under_alice;
```

After executing this we can currently create a view using the "Save as view" button. And then we can sort, apply display formats, filter, encoding...

What I was suggesting is creating a temporary view on the fly and loading that into the table results, instead of the user query. Maybe after a user interaction for "unlocking" sorting and filter features on the table results. Then it should be possible to enable all the features of our Extended Table Widget, shouldn't it? Of course, it would difficult to implement, but probably easier than modifying the select statement directly and re-executing that.

On the other hand, a not so automatic approach, would be simply to allow the users to create views on the temp database, so they can later use them in Data Browse without bothering on when those views will be deleted. It would also allow them to create views in read-only databases. We don't let currently choose the schema in the "Save as view", but it would be useful, I think.

**Comment by MKleusberg** [comment]
@mgrojo Okay I see now. I somehow only thought about creating the view with the desired sorting and then just showing it instead of using the select statement as-is for a view and then sort the view :confused: So you're right, that would work indeed :smile: This also solves the issues I saw about the partial prefetch mechanism, about performance, and about possibly sorting in a different way than the database would.

@sky5walk I agree. @mgrojo made a very similar suggestion but it only creates a view instead of a table, so there is no need to copy all the data (which could take a lot of time for large tables). But the features and therefore the results would be the same.

**Comment by sky5walk** [comment]
Yes, a View is fine as long as you don't limit the rows shown per the preference setting - prefetch block size.
That burned me once or twice.

**Comment by mgrojo** [comment]
@MKleusberg Perfect. Then we have already ideas for enhancement after the release of 3.11.

@sky5walk I think that is already fixed with the last multithreading patch. Is there a description in some issue just for verifying?

**Comment by justinclift** [comment]
Hmmm... I wonder if `CREATE VIEW foo AS ...` can go inside a transaction?  If so, we might be able to automatically put the results of "Execute SQL" queries into an open transaction.

That'd probably isolate them from all other query results (the correct thing to do?), and it'd make sure they don't stick around after the application is closed.

On the other hand though, if the user is trying to change stuff in their database (eg `UPDATE ...`), they probably do want to affect the existing tables.  So not all statements would be suitable for wrapping like that.

**Comment by sky5walk** [comment]
Use [CREATE VIEW TEMP blah...](https://www.sqlite.org/lang_createview.html) and it is dropped when db is closed.

@mrgrojo - I will verify.

**Comment by tlhackque** [comment]
As I noted elsewhere, don't forget read-only databases - you can't "create view" there.

I was initially surprised to find that "create temp view" seems to succeed on a read-only database. But as it's just shorthand for a (stored/named and arbitrarily complex) SELECT statement, it makes sense.

Seems like a solution.

Update should work with a view.  I don't think you need a transaction.  (And it seems from other issues that you wrap everything in a transaction already - to support "write/revert changes".  To nest transactions, you'd have to use SAVEPOINT/RELEASE - which have their own complexities.)
